### ISSUE TITLE
NRE on docking resulting in rapid vessel dissassembly

### ISSUE THREAD

**Issue by darchambault** [jrossignol/ContractConfigurator#410]
I've experienced a rather impressive NRE today, when attempting to dock an Habitat module with an existing station (for the *ExtraCrewCapacity* contract from the "Bases and Stations" contract pack). It resulted in complete and rapid vessel dissassembly, but luckily I had quicksaved seconds before docking.

It's worth mentioning that I somehow received (and accepted) this contract twice (a few in-game days apart), so by docking this station I was fulfilling not one, but two *ExtraCrewCapacity* contracts.

It is easily reproducable from my save file, and I have a complete log file available. The interesting tidibit is here:

```
[LOG 20:37:38.083] Docking to vessel Minmus Station
[LOG 20:37:38.089] Awarding 223404.193550086 funds to player for contract completion
[LOG 20:37:38.090] Awarding 4.270922 reputation to player for contract completion
[LOG 20:37:38.090] Added 2.227988 (4.270922) reputation: 'ContractReward'.
[LOG 20:37:38.092] Contract (Upgrade the crew capacity of Minmus Station): Well done! Bob Kerman would like to thank you. Their nose thanks you too!

<b><#8BED8B>Completion Rewards:</></> 
<#B4D455>Â£223,404   </>
<#E0D503>Â¡4   </>


[EXC 20:37:38.101] ArgumentOutOfRangeException: Argument is out of range.
Parameter name: index
	System.Collections.Generic.List`1[EventData`1+[GameEvents+HostTargetAction`2[Part,Part]]].get_Item (Int32 index)
	EventData`1[GameEvents+HostTargetAction`2[Part,Part]].Fire (HostTargetAction`2 data)
	Part.setParent (.Part p)
	Part.Couple (.Part tgtPart)
	ModuleDockingNode.DockToVessel (.ModuleDockingNode node)
	ModuleDockingNode.<SetupFSM>m__170 ()
	KerbalFSM.RunEvent (.KFSMEvent evt)
	KerbalFSM.updateFSM (KFSMUpdateMode mode)
	KerbalFSM.UpdateFSM ()
	ModuleDockingNode.Update ()
[EXC 20:37:38.109] NullReferenceException: Object reference not set to an instance of an object
	ContractConfigurator.ContractPreLoader+<ContractGenerator>d__12.MoveNext ()
	ContractConfigurator.ContractPreLoader+<ContractGenerator>d__5.MoveNext ()
	ContractConfigurator.ContractPreLoader.GetNextContract (ContractPrestige prestige, Boolean timeLimited)
	ContractConfigurator.ContractPreLoader.GenerateContract (ContractConfigurator.ConfiguredContract contract)
	ContractConfigurator.ConfiguredContract.MeetRequirements ()
	Contracts.Contract.Generate (System.Type contractType, ContractPrestige difficulty, Int32 seed, State state)
	Contracts.ContractSystem.GenerateContract (Int32 seed, ContractPrestige difficulty, System.Type contractType)
	Contracts.ContractSystem.GenerateContract (System.Int32& seed, ContractPrestige difficulty)
	Contracts.ContractSystem.GenerateContracts (System.Int32& seed, ContractPrestige difficulty, Int32 count)
	Contracts.ContractSystem.RefreshContracts ()
	Contracts.ContractSystem+
	.MoveNext ()
[LOG 20:37:38.189] Docking to vessel Minmus Station
[LOG 20:37:38.190] Awarding 217420.577846129 funds to player for contract completion
[LOG 20:37:38.190] Awarding 4.270922 reputation to player for contract completion
[LOG 20:37:38.191] Added 2.203746 (4.270922) reputation: 'ContractReward'.
[LOG 20:37:38.223] Contract (Upgrade the crew capacity of Minmus Station): Well done! Bob Kerman would like to thank you. Their nose thanks you too!

<b><#8BED8B>Completion Rewards:</></> 
<#B4D455>Â£217,421   </>
<#E0D503>Â¡4   </>


[EXC 20:37:38.232] ArgumentOutOfRangeException: Argument is out of range.
Parameter name: index
	System.Collections.Generic.List`1[EventData`1+[GameEvents+HostTargetAction`2[Part,Part]]].get_Item (Int32 index)
	EventData`1[GameEvents+HostTargetAction`2[Part,Part]].Fire (HostTargetAction`2 data)
	Part.setParent (.Part p)
	Part.Couple (.Part tgtPart)
	ModuleDockingNode.DockToVessel (.ModuleDockingNode node)
	ModuleDockingNode.<SetupFSM>m__170 ()
	KerbalFSM.RunEvent (.KFSMEvent evt)
	KerbalFSM.updateFSM (KFSMUpdateMode mode)
	KerbalFSM.UpdateFSM ()
	ModuleDockingNode.Update ()
[LOG 20:37:38.280] [01:31:36]: Structural failure on linkage between Universal Storage: Science Bay and Universal Storage: Octocore.
[LOG 20:37:38.295] [01:31:36]: Structural failure on linkage between Univ. Storage - Soil Moisture Sensor and Universal Storage: Octocore.
[LOG 20:37:38.306] [01:31:36]: Structural failure on linkage between Universal Storage: M30 Monoprop Tank and Universal Storage: Octocore.
[LOG 20:37:38.357] [01:31:36]: Structural failure on linkage between Universal Storage: Octocore and Rockomax Brand Adapter.
```

This is followed by a long list of structural failures :)

Interestingly, the problem disappears if I switch back to the KSC and cancel both *ExtraCrewCapacity* contracts. I've tested cancelling only one of the two contracts, to no effect -- cancelling both is required to avoid the NRE.

I'm attaching:
- the *quicksave.sfs* file
- the *KSP.log* file
- an *installed_mods.ckan* file, for a list of my installed mods

[files.zip](https://github.com/jrossignol/ContractConfigurator/files/70396/files.zip)

**Comment by jrossignol** [comment]
Thanks for the detailed bug report.

I may have a write a workaround for this specific case, but this is a stock bug.  Here's what's happening:
 1. The docking occurs, which causes one of the docking events to fire (not sure which one, probably GameEvents.onPartCouple).
 1. The handler for one of the Contract Configurator parameters fires, and decides that you've completed the contract by docking.  The contract goes away, and all the parameters unregister their event handlers.
 1. The docking event tries to fire the next event handler, but because we removed a whole bunch from the list (each parameter is going to be registering events handlers), it's not pointing at an invalid spot in the list and falls over.


@NathanKell - is there any chance you can fix this one for 1.1?  The fix should be relatively straightforward:

 1. In EventData<T> through EventData<T, U, V, W>, the Fire() method needs a conditional that if the index is >= events.Count, then set it to events.Count - 1
 1. If the new index is less than 0, break or bad things happen.

Let me know if you want me to tracker this somewhere, or if you want to discuss on other channels.

**Comment by darchambault** [comment]
Thanks for the reply and the explanation!

Just to clarify, for my own curiosity, this happens when two contracts are registered to the same "event", like in my case both contract being triggered by the docking event on the same station?

If so, wouldn't cancelling one of the two contracts be sufficient to resolve the issue? Because I tried that with no success.

**Comment by jrossignol** [comment]
It's actually parameters that are registering for events, and most Contract Configurator contracts have many, *many* parameters.  So it's quite easy to get this situation with a single contract.

**Comment by darchambault** [comment]
Does that mean the *ExtraCrewCapacity* contract from "Bases and Stations" is broken, and I should keep from accepting contracts of that type for now?

**Comment by jrossignol** [comment]
It's just *mostly* broken - if you complete the contract without docking (EVA the kerbals over), it'll probably work.  Oh, and you're likely getting two duplicates because of #402.

**Comment by darchambault** [comment]
I don't think EVA is an option, I need to add crew _capacity_ to complete the contract, not actual crew :) I think I'll have to cancel the contracts for now, dock, and carry on.

**Comment by NathanKell** [comment]
Yep, YANFRET also PM'd me about this (in re Chaka). Apparently there is an issue on the tracker which sal has confirmed. Definitely will be fixing it for 1.1 (but do add it to the current tracker @jrossignol or copy the one from the public tracker? )

**Comment by severedsolo** [comment]
You'll need to add some extra space to your station (with or without the contract) to stop them from generating for now, it's looking for any stations with less than 4 (maybe 5, but I'm 99% sure it's 4) empty seats.

**Comment by jrossignol** [comment]
Confirmed as fixed in 1.1.

**Comment by jrossignol** [comment]
Fixed by KSP 1.1
